File: app/src/main/java/io/github/finall1008/xiaoaimcp/ui/ServerEditActivity.java
package io.github.finall1008.xiaoaimcp.ui;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.InputType;
import android.text.method.PasswordTransformationMethod;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.Spinner;
import android.widget.Switch;
import android.widget.TextView;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;

import io.github.finall1008.xiaoaimcp.BridgeApplication;
import io.github.finall1008.xiaoaimcp.config.McpConfigValidator;
import io.github.finall1008.xiaoaimcp.config.McpServer;
import io.github.finall1008.xiaoaimcp.config.RemoteConfigRepository;

public final class ServerEditActivity extends Activity {
    public static final String EXTRA_SERVER_ID = "server_id";

    private String serverId;
    private EditText nameInput;
    private EditText descriptionInput;
    private EditText urlInput;
    private EditText headersInput;
    private Spinner transportInput;
    private Switch enabledInput;
    private CheckBox showHeadersInput;
    private RemoteConfigRepository repository;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        serverId = getIntent().getStringExtra(EXTRA_SERVER_ID);
        SharedPreferences preferences = BridgeApplication.remotePreferences();
        if (preferences == null) {
            new AlertDialog.Builder(this)
                    .setTitle("API 102 服务未连接")
                    .setMessage("无法读取或保存服务器配置。")
                    .setPositiveButton("关闭", (dialog, which) -> finish())
                    .setCancelable(false)
                    .show();
            return;
        }
        repository = new RemoteConfigRepository(preferences);
        buildUi();
        if (serverId != null) {
            loadExisting();
        }
    }

    private void buildUi() {
        ScrollView scroll = new ScrollView(this);
        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setPadding(UiSupport.dp(this, 20), UiSupport.dp(this, 20),
                UiSupport.dp(this, 20), UiSupport.dp(this, 28));
        scroll.addView(root, UiSupport.matchWrap());

        root.addView(UiSupport.title(this,
                serverId == null ? "添加 MCP 服务器" : "编辑 MCP 服务器"), UiSupport.matchWrap());
        TextView hint = UiSupport.text(this,
                "HTTP 对应 Streamable HTTP；SSE 对应旧式 Server-Sent Events。配置保存后会通知超级小爱在线重载。",
                13);
        root.addView(hint, UiSupport.matchWrap());

        nameInput = UiSupport.labeledEdit(root, "名称", "例如 my-search");
        descriptionInput = UiSupport.labeledEdit(root, "说明（可选）", "此服务器提供什么能力");
        urlInput = UiSupport.labeledEdit(root, "服务器 URL", "https://example.com/mcp");
        urlInput.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_URI);

        TextView transportLabel = UiSupport.text(this, "传输类型", 13);
        transportLabel.setPadding(0, UiSupport.dp(this, 12), 0, 0);
        root.addView(transportLabel, UiSupport.matchWrap());
        transportInput = new Spinner(this);
        transportInput.setAdapter(new ArrayAdapter<>(this,
                android.R.layout.simple_spinner_dropdown_item,
                new String[]{"http", "sse"}));
        root.addView(transportInput, UiSupport.matchWrap());

        enabledInput = new Switch(this);
        enabledInput.setText("启用服务器");
        enabledInput.setChecked(true);
        enabledInput.setPadding(0, UiSupport.dp(this, 12), 0, UiSupport.dp(this, 6));
        root.addView(enabledInput, UiSupport.matchWrap());

        headersInput = UiSupport.labeledEdit(root, "请求头（每行 Name: Value）",
                "Authorization: Bearer token\nX-API-Key: secret");
        headersInput.setSingleLine(false);
        headersInput.setMinLines(4);
        headersInput.setGravity(android.view.Gravity.TOP);
        headersInput.setInputType(InputType.TYPE_CLASS_TEXT
                | InputType.TYPE_TEXT_FLAG_MULTI_LINE
                | InputType.TYPE_TEXT_VARIATION_PASSWORD);
        headersInput.setTransformationMethod(PasswordTransformationMethod.getInstance());

        showHeadersInput = new CheckBox(this);
        showHeadersInput.setText("显示请求头值");
        showHeadersInput.setOnCheckedChangeListener((button, checked) -> {
            int selection = headersInput.getSelectionStart();
            headersInput.setTransformationMethod(
                    checked ? null : PasswordTransformationMethod.getInstance());
            headersInput.setSelection(Math.max(0, selection));
        });
        root.addView(showHeadersInput, UiSupport.matchWrap());

        Button save = new Button(this);
        save.setText("保存并应用");
        save.setAllCaps(false);
        save.setOnClickListener(v -> save());
        LinearLayout.LayoutParams saveParams = new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
        );
        saveParams.topMargin = UiSupport.dp(this, 16);
        root.addView(save, saveParams);
        setContentView(scroll);
    }

    private void loadExisting() {
        try {
            McpServer server = repository.findById(serverId);
            if (server == null) {
                throw new IllegalStateException("服务器不存在或已删除");
            }
            nameInput.setText(server.name());
            descriptionInput.setText(server.description());
            urlInput.setText(server.url());
            transportInput.setSelection(server.transport().equals("sse") ? 1 : 0);
            enabledInput.setChecked(server.enabled());
            headersInput.setText(formatHeaders(server.headers()));
        } catch (Exception error) {
            showError(error);
        }
    }

    private void save() {
        try {
            McpServer server = new McpServer(
                    serverId == null ? UUID.randomUUID().toString() : serverId,
                    nameInput.getText().toString(),
                    descriptionInput.getText().toString(),
                    urlInput.getText().toString(),
                    transportInput.getSelectedItem().toString(),
                    enabledInput.isChecked(),
                    parseHeaders(headersInput.getText().toString())
            );
            McpConfigValidator.validate(server);
            repository.upsert(server);
            finish();
        } catch (Exception error) {
            showError(error);
        }
    }

    static Map<String, String> parseHeaders(String text) throws Exception {
        Map<String, String> headers = new LinkedHashMap<>();
        String normalized = text == null ? "" : text.replace("\r\n", "\n");
        for (String rawLine : normalized.split("\n")) {
            String line = rawLine.trim();
            if (line.isEmpty()) {
                continue;
            }
            int separator = line.indexOf(':');
            if (separator <= 0) {
                throw new IllegalArgumentException("请求头每行必须使用 Name: Value 格式");
            }
            String name = line.substring(0, separator).trim();
            String value = line.substring(separator + 1).trim();
            if (headers.containsKey(name)) {
                throw new IllegalArgumentException("请求头名称重复：" + name);
            }
            headers.put(name, value);
        }
        return headers;
    }

    private static String formatHeaders(Map<String, String> headers) {
        StringBuilder builder = new StringBuilder();
        for (Map.Entry<String, String> entry : headers.entrySet()) {
            if (builder.length() > 0) {
                builder.append('\n');
            }
            builder.append(entry.getKey()).append(": ").append(entry.getValue());
        }
        return builder.toString();
    }

    private void showError(Throwable error) {
        String message = error.getMessage();
        new AlertDialog.Builder(this)
                .setTitle("无法保存")
                .setMessage(message == null || message.isBlank()
                        ? error.getClass().getSimpleName() : message)
                .setPositiveButton("确定", null)
                .show();
    }
}

File: app/src/main/java/io/github/finall1008/xiaoaimcp/ui/ServerEditActivity.kt
package io.github.finall1008.xiaoaimcp.ui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.unit.dp
import io.github.finall1008.xiaoaimcp.BridgeApplication
import io.github.finall1008.xiaoaimcp.config.McpConfigValidator
import io.github.finall1008.xiaoaimcp.config.McpServer
import io.github.finall1008.xiaoaimcp.config.RemoteConfigRepository
import top.yukonga.miuix.kmp.basic.Card
import top.yukonga.miuix.kmp.basic.Text
import top.yukonga.miuix.kmp.basic.TextField
import top.yukonga.miuix.kmp.preference.CheckboxLocation
import top.yukonga.miuix.kmp.preference.CheckboxPreference
import top.yukonga.miuix.kmp.preference.OverlayDropdownPreference
import top.yukonga.miuix.kmp.preference.SwitchPreference
import top.yukonga.miuix.kmp.theme.MiuixTheme
import java.util.LinkedHashMap
import java.util.UUID

class ServerEditActivity : ComponentActivity() {
    private val serverId: String? by lazy { intent.getStringExtra(EXTRA_SERVER_ID) }
    private var repository: RemoteConfigRepository? = null
    private var form by mutableStateOf(ServerForm())
    private var errorMessage by mutableStateOf<String?>(null)
    private var fatalMessage by mutableStateOf<String?>(null)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val preferences = BridgeApplication.remotePreferences()
        if (preferences == null) {
            fatalMessage = "无法读取或保存服务器配置。"
        } else {
            repository = RemoteConfigRepository(preferences)
            if (serverId != null) loadExisting()
        }
        setContent {
            BridgeTheme(this) {
                ServerEditScreen(
                    editing = serverId != null,
                    form = form,
                    onFormChange = { form = it },
                    onBack = ::finish,
                    onSave = ::save,
                    errorMessage = errorMessage,
                    onDismissError = { errorMessage = null },
                    fatalMessage = fatalMessage,
                    onDismissFatal = ::finish,
                )
            }
        }
    }

    private fun loadExisting() {
        try {
            val server = repository?.findById(serverId)
                ?: error("服务器不存在或已删除")
            form = ServerForm(
                name = server.name(),
                description = server.description(),
                url = server.url(),
                transportIndex = if (server.transport() == "sse") 1 else 0,
                enabled = server.enabled(),
                headers = formatHeaders(server.headers()),
            )
        } catch (error: Exception) {
            errorMessage = safeMessage(error)
        }
    }

    private fun save() {
        try {
            val server = McpServer(
                serverId ?: UUID.randomUUID().toString(),
                form.name,
                form.description,
                form.url,
                TRANSPORTS[form.transportIndex],
                form.enabled,
                parseHeaders(form.headers),
            )
            McpConfigValidator.validate(server)
            (repository ?: error("API 102 服务未连接")).upsert(server)
            finish()
        } catch (error: Exception) {
            errorMessage = safeMessage(error)
        }
    }

    companion object {
        const val EXTRA_SERVER_ID = "server_id"
        private val TRANSPORTS = listOf("http", "sse")

        @JvmStatic
        internal fun parseHeaders(text: String?): Map<String, String> {
            val headers = LinkedHashMap<String, String>()
            val normalized = text.orEmpty().replace("\r\n", "\n")
            normalized.split('\n').forEach { rawLine ->
                val line = rawLine.trim()
                if (line.isEmpty()) return@forEach
                val separator = line.indexOf(':')
                require(separator > 0) { "请求头每行必须使用 Name: Value 格式" }
                val name = line.substring(0, separator).trim()
                val value = line.substring(separator + 1).trim()
                require(!headers.containsKey(name)) { "请求头名称重复：$name" }
                headers[name] = value
            }
            return headers
        }

        private fun formatHeaders(headers: Map<String, String>): String {
            return headers.entries.joinToString("\n") { "${it.key}: ${it.value}" }
        }
    }
}

private data class ServerForm(
    val name: String = "",
    val description: String = "",
    val url: String = "",
    val transportIndex: Int = 0,
    val enabled: Boolean = true,
    val headers: String = "",
    val showHeaders: Boolean = false,
)

@Composable
private fun ServerEditScreen(
    editing: Boolean,
    form: ServerForm,
    onFormChange: (ServerForm) -> Unit,
    onBack: () -> Unit,
    onSave: () -> Unit,
    errorMessage: String?,
    onDismissError: () -> Unit,
    fatalMessage: String?,
    onDismissFatal: () -> Unit,
) {
    BridgePageScaffold(
        title = if (editing) "编辑 MCP 服务器" else "添加 MCP 服务器",
        onBack = onBack,
        bottomBar = {
            BridgeSaveBar(
                text = "保存并应用",
                enabled = fatalMessage == null,
                onClick = onSave,
            )
        },
    ) { padding, scrollBehavior ->
        BridgePageList(
            scaffoldPadding = padding,
            scrollBehavior = scrollBehavior,
        ) {
            item { SectionLabel("基本信息") }
            item {
                Card(
                    modifier = Modifier.fillMaxWidth(),
                    insideMargin = PaddingValues(16.dp),
                ) {
                    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
                        TextField(
                            value = form.name,
                            onValueChange = { onFormChange(form.copy(name = it)) },
                            modifier = Modifier.fillMaxWidth(),
                            label = "名称，例如 my-search",
                            singleLine = true,
                        )
                        TextField(
                            value = form.description,
                            onValueChange = { onFormChange(form.copy(description = it)) },
                            modifier = Modifier.fillMaxWidth(),
                            label = "说明（可选）",
                            minLines = 3,
                            maxLines = 6,
                        )
                        TextField(
                            value = form.url,
                            onValueChange = { onFormChange(form.copy(url = it)) },
                            modifier = Modifier.fillMaxWidth(),
                            label = "服务器 URL",
                            singleLine = true,
                            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Uri),
                        )
                    }
                }
            }
            item { SectionLabel("连接设置", Modifier.padding(top = 8.dp)) }
            item {
                Card(modifier = Modifier.fillMaxWidth()) {
                    OverlayDropdownPreference(
                        items = listOf("http", "sse"),
                        selectedIndex = form.transportIndex,
                        title = "传输类型",
                        summary = "HTTP 为 Streamable HTTP；SSE 为旧式连接",
                        onSelectedIndexChange = {
                            onFormChange(form.copy(transportIndex = it))
                        },
                    )
                    SwitchPreference(
                        checked = form.enabled,
                        onCheckedChange = { onFormChange(form.copy(enabled = it)) },
                        title = "启用服务器",
                        summary = "关闭后保留配置，但不会向超级小爱注册工具",
                    )
                }
            }
            item { SectionLabel("请求头", Modifier.padding(top = 8.dp)) }
            item {
                Card(
                    modifier = Modifier.fillMaxWidth(),
                    insideMargin = PaddingValues(16.dp),
                ) {
                    Text(
                        "每行使用 Name: Value 格式",
                        style = MiuixTheme.textStyles.footnote1,
                        color = MiuixTheme.colorScheme.onSurfaceVariantSummary,
                    )
                    Spacer(Modifier.height(10.dp))
                    TextField(
                        value = form.headers,
                        onValueChange = { onFormChange(form.copy(headers = it)) },
                        modifier = Modifier.fillMaxWidth(),
                        label = "Authorization: Bearer token\nX-API-Key: secret",
                        minLines = 4,
                        maxLines = 8,
                        visualTransformation = if (form.showHeaders) {
                            VisualTransformation.None
                        } else {
                            PasswordVisualTransformation()
                        },
                    )
                    CheckboxPreference(
                        title = "显示请求头值",
                        checked = form.showHeaders,
                        onCheckedChange = { onFormChange(form.copy(showHeaders = it)) },
                        checkboxLocation = CheckboxLocation.End,
                    )
                }
            }
        }
        MessageDialog(
            message = errorMessage,
            title = "无法保存",
            onDismiss = onDismissError,
        )
        MessageDialog(
            message = fatalMessage,
            title = "API 102 服务未连接",
            onDismiss = onDismissFatal,
        )
    }
}
